caps/management/commands/process_scottish_reporting_data.py

```diff
@@ -0,0 +1,428 @@
+            "start_text": "Reference year",
+        },
+        "sources": {
+            "desc": "emissions sources data",
+            "method": "extract_emissions_sources_from_sheet",
+            "header_text": "Emission source",
+        },
+        "targets": {
+            "desc": "reduction targets data",
+            "method": "extract_targets_from_sheet",
+            "header_text": r"Name of [Tt]arget",
+        },
+        "projects": {
+            "desc": "emission reduction projects data",
+            "method": "extract_projects_from_sheet",
+            "header_text": "Project name",
+        },
+        "generation": {
+            "desc": "renewables generation data",
+            "method": "extract_generation_from_sheet",
+            "header_text": "Generation, consumption and export of renewable energy",
+            "start_text": "Technology",
+        },
+    }
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            "--all",
+            default=True,
+            action="store_true",
+            help="Update all data",
+        )
+
+        parser.add_argument(
+            "--sample",
+            action="store_true",
+            help="Just process the first 20 rows for testing",
+        )
+
+        for arg, desc in self.subsets.items():
+            parser.add_argument(
+                "--{}".format(arg),
+                action="store_true",
+                help="Update {}".format(desc),
+            )
+
+    def handle(self, *args, **options):
+        self.options = options
+
+        get_all = options["all"]
+        for option in self.subsets.keys():
+            if options[option]:
+                self.options["all"] = False
+                break
+
+        self.process_data()
+        self.save_data()
+
+    def process_data(self):
+        data = join(settings.SCOTTISH_DIR, "data_list.csv")
+        df = pd.read_csv(data)
+
+        for index, row in df.iterrows():
+            try:
+                self.report_data = row
+                self.process_report()
+                if self.options["sample"] and index == 20:
+                    break
+            except Exception as e:
+                print(
+                    "problem processing report {} for council {}: {}".format(
+                        row["file"], row["council"], e
+                    )
+                )
+
+    def process_report(self):
+        # if data["start_year"] != 2020 and data["start_year"] != 2021:
+        # return
+        if pd.isna(self.report_data["file"]):
+            print(
+                "bad file for {} {}".format(
+                    self.report_data["council"], self.report_data["start_year"]
+                )
+            )
+            return
+
+        self.report = join(settings.SCOTTISH_DIR, self.report_data["file"])
+
+        xls = pd.ExcelFile(self.report)
+        parsed = {}
+        for sheet_name in xls.sheet_names:
+            self.sheet_name = sheet_name
```

Comment:
Think it is working fine here, but this repeated assignment to a self makes me nervous in case something is missed and it comes around on next loop. If re-factoring at some point would probably move the sheet processing into its own class that can be loaded and destroyed?